test/jdk/java/lang/invoke/common/test/java/lang/invoke/lib/InstructionHelper.java

```diff
@@ -23,121 +23,104 @@
+                                        .aload(0)
+                                        .invokespecial(ConstantDescs.CD_Object, ConstantDescs.INIT_NAME,
+                                                ConstantDescs.MTD_void, false)
+                                        .return_()));
     }
 
-    public static MethodHandle ldcDynamicConstant(MethodHandles.Lookup l,
-                                                  String name, Class<?> type,
-                                                  String bsmMethodName, MethodType bsmType,
-                                                  Consumer<PoolHelper.StaticArgListBuilder<String, String, byte[]>> staticArgs) throws Exception {
-        return ldcDynamicConstant(l, name, type, l.lookupClass(), bsmMethodName, bsmType, staticArgs);
+    private static String className(MethodHandles.Lookup l) {
+        return l.lookupClass().getCanonicalName().replace('.', '/') + "$Code_" +
+                COUNT.getAndIncrement();
     }
 
-    public static MethodHandle ldcDynamicConstant(MethodHandles.Lookup l,
-                                                  String name, Class<?> type,
-                                                  Class<?> bsmClass, String bsmMethodName, MethodType bsmType,
-                                                  Consumer<PoolHelper.StaticArgListBuilder<String, String, byte[]>> staticArgs) throws Exception {
-        return ldcDynamicConstant(l, name, cref(type), csym(bsmClass), bsmMethodName, bsmType.toMethodDescriptorString(), staticArgs);
+    public static MethodHandle invokedynamic(MethodHandles.Lookup l, String name, MethodType type, String bsmMethodName,
+                                             MethodType bsmType,
+                                             ConstantDesc[] boostrapArgs) throws Exception {
+        byte[] byteArray = Classfile.of().build(ClassDesc.of(className(l)), classBuilder -> {
+            commonBuild(classBuilder);
+            classBuilder
+                    .withMethod("m", MethodTypeDesc.ofDescriptor(type.toMethodDescriptorString()),
+                            Classfile.ACC_PUBLIC + Classfile.ACC_STATIC, methodBuilder -> methodBuilder
+                                    .withCode(codeBuilder -> {
+                                        for (int i = 0; i < type.parameterCount(); i++) {
+                                            codeBuilder.loadInstruction(TypeKind.fromDescriptor(
+                                                    type.parameterType(i).descriptorString()), i);
+                                        }
+                                        codeBuilder.invokedynamic(DynamicCallSiteDesc.of(
+                                                MethodHandleDesc.ofMethod(
+                                                        DirectMethodHandleDesc.Kind.STATIC,
+                                                        ClassDesc.of(l.lookupClass().getCanonicalName()),
```

Comment:
Despite this being part of the original code, `getCanonicalName` is unreliable for bytecode engineering, for it cannot distinguish packages from nested/inner classes.

Reply:
I will use ClassDesc.ofDescriptor and passes the descriptor string of the type.